src/directives/class-map.ts

```diff
@@ -71,13 +73,18 @@ export const classMap = directive((classInfo: ClassInfo) => (part: Part) => {
     // We explicitly want a loose truthy check of `value` because it seems more
     // convenient that '' and 0 are skipped.
     if (value != previousClasses.has(name)) {
+      changed = true;
       if (value) {
-        classList.add(name);
-        previousClasses.add(name);
+        previousClasses.set(name, true);
       } else {
-        classList.remove(name);
         previousClasses.delete(name);
       }
     }
   }
-});
+
+  if (changed) {
+    const classes: string[] = [];
+    previousClasses.forEach((_: unknown, key: string) => classes.push(key));
+    element.setAttribute('class', classes.join(' '));
```

Comment:
Since we're not using `.classList` anymore, this directive is essentially just doing custom dirty checking, and joining the property names of the `classInfo` object. I think there's a slightly simpler approach by using `part.setValue`.

Take the version at master, rip out the use of `classList` and setting of `className`, keep a `changed` variable as you do here, but then do:

```ts
if (changed) {
  part.setValue(Array.from(previousClasses).join(' ');
}
```

The AttributePart logic will take care of concatenating that with the static strings, so you won't need to keep them in the `previousClasses` Map.

Reply:
> The AttributePart logic will take care of concatenating that with the static strings, so you won't need to keep them in the previousClasses Map.

That would be a breaking change, since you can currently override static classes with the `ClassInfo`. Otherwise, yes, it would be simpler.